Switch to multi-stage build

mostafahussein opened the pull request:
- Dockerfile updated to use [multi-stage builds](https://docs.docker.com/develop/develop-images/multistage-build/).
- Unnecessary commands in Dockerfile removed.
- NodeGoat image size reduced from 951 MB to 97.4 MB by using `node:4-alpine`
> If the image is affecting the project let me know so i can revert it to `node:4.4` as the version of node:4-alpine currently is 4.9.1 and the version of node:4.4 is 4.4.7 however i did not consider it a problem as the documentation says: requires Node v4.4 or above
- Removed `links` from docker-compose.yml as by default the 2 services will be linked with the network that will be created when you run `docker-compose up`

lirantal commented:
Looks ok to me.
Do we have a reason not to use alpine, such as any demos this might impact?

mostafahussein commented:
Generally speaking, using alpine based or debian/centos based most of time it was related to `npm install` process (the packages that needs to be built during the installation)  but as we have it in a separate stage it wont be a problem.

Regarding NodeGoat in specific, it does not mention any specific distribution to use, so i assume that using alpine or debian based wont be an issue unless you have a certain demo you want to test to make sure of the stability of the project then let me know which one so i can take a look at it.

binarymist commented:
Hi @mostafahussein

Can you copy->past the file ownership and permissions please? Does the file ownership and permissions still look like the second example [here](https://f1.holisticinfosecforwebdevelopers.com/chap03.html#vps-countermeasures-docker-the-default-user-is-root)?

The:

```
RUN ls -liah  
RUN ls ../ -liah
```

Is to demonstrate the ownership and permissions changes.

In the `docker-compose.yml` the:

```
    expose: 
      - "27017"
```

Is essentially documentation. Is there a reason to remove it?

`links` expresses the dependency. `mongo` needs to be running first.

binarymist commented:
I also don't see any reason why we shouldn't use apline if using multi-stage builds

mostafahussein commented:
Hi @binarymist ,

- Regarding the output of:
```
RUN ls -liah  
RUN ls ../ -liah
```
The result will showing that `app` directory is owned by `nodegoat_docker` also i suggest to rename it `nodegoat` only for readability because it will displayed as nodegoad only so to avoid confusion naming it to `nodegoat` might be a better option 
Here is the output:
```
Step 14/18 : RUN ls -liah
 ---> Running in a60cb9fd5731
total 508
3813706 drwxr-sr-x    1 nodegoat nodegoat    4.0K Mar 22 21:41 .
3813705 drwxr-sr-x    1 nodegoat nodegoat    4.0K Mar 22 14:28 ..
3813707 drwxr-xr-x    1 nodegoat nodegoat    4.0K Mar 22 15:03 .git
3813778 -rw-r--r--    1 nodegoat nodegoat     178 Mar 22 14:10 .gitignore
3813779 -rw-r--r--    1 nodegoat nodegoat    1.9K Mar 22 14:10 .jshintrc
3813780 -rw-r--r--    1 nodegoat nodegoat      55 Mar 22 14:10 .nodemonignore
3813781 -rw-r--r--    1 nodegoat nodegoat     667 Mar 22 21:41 Dockerfile
3813782 -rw-r--r--    1 nodegoat nodegoat    6.6K Mar 22 14:10 Gruntfile.js
3813783 -rw-r--r--    1 nodegoat nodegoat   10.0K Mar 22 14:10 LICENSE
3813784 -rw-r--r--    1 nodegoat nodegoat      48 Mar 22 14:10 Procfile
3813785 -rw-r--r--    1 nodegoat nodegoat    5.5K Mar 22 14:10 README.md
3813786 drwxr-xr-x    1 nodegoat nodegoat    4.0K Mar 22 14:10 app
3813787 -rw-r--r--    1 nodegoat nodegoat     527 Mar 22 14:10 app.json
3947501 drwxr-xr-x    1 nodegoat nodegoat    4.0K Mar 22 14:10 artifacts
3947506 drwxr-xr-x    1 nodegoat nodegoat    4.0K Mar 22 14:10 config
3813805 -rw-r--r--    1 nodegoat nodegoat     192 Mar 22 14:45 docker-compose.yml
3947513 drwxr-sr-x    1 nodegoat nodegoat    4.0K Mar 22 21:41 node_modules
3813806 -rw-r--r--    1 nodegoat nodegoat  404.5K Mar 22 14:10 package-lock.json
3813807 -rw-r--r--    1 nodegoat nodegoat    1.6K Mar 22 14:10 package.json
3813808 -rw-r--r--    1 nodegoat nodegoat    4.6K Mar 22 14:10 server.js
3813809 drwxr-xr-x    1 nodegoat nodegoat    4.0K Mar 22 14:10 test
Removing intermediate container a60cb9fd5731
 ---> 6101db444b69
Step 15/18 : RUN ls ../ -liah
 ---> Running in 5e8b3631a290
total 12
3813705 drwxr-sr-x    1 nodegoat nodegoat    4.0K Mar 22 14:28 .
3813704 drwxr-xr-x    1 root     root        4.0K Mar 22 14:28 ..
3813706 drwxr-sr-x    1 nodegoat nodegoat    4.0K Mar 22 21:41 app
Removing intermediate container 5e8b3631a290
```
I just removed it to make lower the total of the image layers.

- `expose` existence is not affecting the application functionality because both of `web` and `mongo` are inside the same network. so `web`  will access the mongodb's port normally in this case.

- lastly the `links` indeed represent the dependency and the order as you said however it does not grantee that the mongodb is ready for connections. if that's what you mean then can add a script like `wait-for` which wont run any commands from `web` unless `mongo` is ready for connections, what do you think ?

binarymist commented:
We probably don't need the application files to be world readable, hence the existing `RUN chmod --recursive o-wrx $workdir`, it'd be good to re-include this or something similar that has the same effect.

> I just removed it to make lower the total of the image layers, other than that i didn't know the reason behind adding it so i have removed it.

Removed what (what is "it")?

I think it makes sense to have self documenting code, please leave the `expose` command intact.

> lastly the links indeed represent the dependency and the order as you said however it does not guarantee that the mongodb is ready for connections. if that's what you mean then can add a script like wait-for which wont run any commands from web unless mongo is ready for connections, what do you think ?

Sounds good, a `wait-for` script is a good idea.

Thanks.

mostafahussein commented:
I have added `wait-for` and the other modifications requested, also i have updated my previous comment regarding `it` to make it clearer.

The output is like below:
```
Step 14/17 : RUN ls -liah
 ---> Running in ca794db3d76a
total 508
4456536 drwxr-s---    1 nodegoat nodegoat    4.0K Mar 22 23:48 .
4456535 drwxr-sr-x    1 nodegoat nodegoat    4.0K Mar 22 14:28 ..
4456537 drwxr-x---    1 nodegoat nodegoat    4.0K Mar 22 23:47 .git
4456608 -rw-r-----    1 nodegoat nodegoat     178 Mar 22 14:10 .gitignore
4456609 -rw-r-----    1 nodegoat nodegoat    1.9K Mar 22 14:10 .jshintrc
4456610 -rw-r-----    1 nodegoat nodegoat      55 Mar 22 14:10 .nodemonignore
4456611 -rw-r-----    1 nodegoat nodegoat     674 Mar 22 23:48 Dockerfile
4456612 -rw-r-----    1 nodegoat nodegoat    6.6K Mar 22 14:10 Gruntfile.js
4456613 -rw-r-----    1 nodegoat nodegoat   10.0K Mar 22 14:10 LICENSE
4456614 -rw-r-----    1 nodegoat nodegoat      48 Mar 22 14:10 Procfile
4456615 -rw-r-----    1 nodegoat nodegoat    5.5K Mar 22 14:10 README.md
4456616 drwxr-x---    1 nodegoat nodegoat    4.0K Mar 22 14:10 app
4456617 -rw-r-----    1 nodegoat nodegoat     527 Mar 22 14:10 app.json
4456693 drwxr-x---    1 nodegoat nodegoat    4.0K Mar 22 14:10 artifacts
4456698 drwxr-x---    1 nodegoat nodegoat    4.0K Mar 22 14:10 config
4461760 -rw-r-----    1 nodegoat nodegoat     218 Mar 22 23:46 docker-compose.yml
4461761 drwxr-s---    1 nodegoat nodegoat    4.0K Mar 22 23:48 node_modules
4464755 -rw-r-----    1 nodegoat nodegoat  404.5K Mar 22 14:10 package-lock.json
4464756 -rw-r-----    1 nodegoat nodegoat    1.6K Mar 22 14:10 package.json
4464757 -rw-r-----    1 nodegoat nodegoat    4.6K Mar 22 14:10 server.js
4464758 drwxr-x---    1 nodegoat nodegoat    4.0K Mar 22 14:10 test
Removing intermediate container ca794db3d76a
 ---> 07b5d8c8a099
Step 15/17 : RUN ls ../ -liah
 ---> Running in e20254d0c066
total 12
4456535 drwxr-sr-x    1 nodegoat nodegoat    4.0K Mar 22 14:28 .
4456534 drwxr-xr-x    1 root     root        4.0K Mar 22 14:28 ..
4456536 drwxr-s---    1 nodegoat nodegoat    4.0K Mar 22 23:48 app
Removing intermediate container e20254d0c066
```
What else needs to be modified ?

binarymist commented:
May want to make the user you're creating a system user (`-S`). I use this option [here](https://gitlab.com/purpleteam-labs/purpleteam-orchestrator/blob/master/Dockerfile#L20) for example.

I'm not sure how you're getting `nodegoat` as the user, group listing when `USER` is still defined as `nodegoat_docker`?

Have you tested it?

Thanks.

mostafahussein commented:
I have modified the Dockerfile to create system user/group (I haven't pushed the changes yet). Regarding the name in ls output i am not sure also why it appears like this but maybe something related with the name length as the image is works as expected and yes i have tested it by running the application and log in, I have checked `/etc/passwd` and the user is there as expected. What about renaming it to `nodegoat` instead ?

mostafahussein commented:
- I found that there is `s` permission for group - _as you will notice in the output in my previous comment_ - added in the second stage i have added: `chmod -R g-s /home/$USER` to remove this extra permission.

- I have added `.dockerignore` file and added the following content to it in order to avoid having these files in the final image:
```
Dockerfile
docker-compose.yml
.dockerignore
.git
.gitignore
```
The current output as below:
```
Step 16/19 : RUN ls -liah
 ---> Running in 11108f861422
total 492
10362009 drwxr-x---    1 nodegoat nodegoat    4.0K Mar 23 11:29 .
10362008 drwxr-xr-x    1 nodegoat nodegoat    4.0K Mar 23 09:13 ..
10362010 -rw-r-----    1 nodegoat nodegoat    1.9K Mar 22 14:10 .jshintrc
10362011 -rw-r-----    1 nodegoat nodegoat      55 Mar 22 14:10 .nodemonignore
10362012 -rw-r-----    1 nodegoat nodegoat    6.6K Mar 22 14:10 Gruntfile.js
10362013 -rw-r-----    1 nodegoat nodegoat   10.0K Mar 22 14:10 LICENSE
10362014 -rw-r-----    1 nodegoat nodegoat      48 Mar 22 14:10 Procfile
10362015 -rw-r-----    1 nodegoat nodegoat    5.5K Mar 22 14:10 README.md
10362016 drwxr-x---    1 nodegoat nodegoat    4.0K Mar 22 14:10 app
10362017 -rw-r-----    1 nodegoat nodegoat     527 Mar 22 14:10 app.json
10362094 drwxr-x---    1 nodegoat nodegoat    4.0K Mar 22 14:10 artifacts
10362099 drwxr-x---    1 nodegoat nodegoat    4.0K Mar 22 14:10 config
10362106 drwxr-x---    1 nodegoat nodegoat    4.0K Mar 23 11:29 node_modules
10362242 -rw-r-----    1 nodegoat nodegoat  404.5K Mar 22 14:10 package-lock.json
10362243 -rw-r-----    1 nodegoat nodegoat    1.6K Mar 22 14:10 package.json
10362244 -rw-r-----    1 nodegoat nodegoat    4.6K Mar 22 14:10 server.js
10362245 drwxr-x---    1 nodegoat nodegoat    4.0K Mar 22 14:10 test
Removing intermediate container 11108f861422
 ---> 9b4474464ed1
Step 17/19 : RUN ls ../ -liah
 ---> Running in fea7bf92c662
total 12
10362008 drwxr-xr-x    1 nodegoat nodegoat    4.0K Mar 23 09:13 .
10362007 drwxr-xr-x    1 root     root        4.0K Mar 23 09:13 ..
10362009 drwxr-x---    1 nodegoat nodegoat    4.0K Mar 23 11:29 app
Removing intermediate container fea7bf92c662
```

ckarande commented:
@mostafahussein @lirantal @binarymist thanks for collaborating on it.  I believe the only open item is whether to rename owner `nodegoat_docker` to `nodegoat`. @binarymist please suggest if you have any preferences on that? If we are all set, I will be happy to test it once before we merge it. Cheers!

binarymist commented:
I have no strong preference over `nodegoat_docker` vs `nodegoat`.

---

In order to take advantage of Docker caching, it might be an idea to move the `COPY . $WORKDIR` to as late as practical, probably after the following:

```
ADD https://raw.githubusercontent.com/eficode/wait-for/master/wait-for /usr/local/bin/wait-for
```

---

Any other tweaks we should be able to submit another PR if required.

Thanks.

mostafahussein commented:
- For the username I will change it to nodegoat for simplicity.
- For Docker caching i don't see the difference if we changed the place of `COPY . $WORKDIR`. The current is showing as below as you can see the cache is working:
```
Step 11/17 : WORKDIR $WORKDIR
 ---> Using cache
 ---> 94142953f6e2
Step 12/17 : COPY . $WORKDIR
 ---> Using cache
 ---> c2ffcaa47b1c
Step 13/17 : COPY --from=0 /usr/src/app/node_modules node_modules
 ---> Using cache
 ---> c6e5a8ce1b8c
Step 14/17 : ADD https://raw.githubusercontent.com/eficode/wait-for/master/wait-for /usr/local/bin/wait-for
Downloading  1.378kB
 ---> Using cache
 ---> 339b3e249996
Step 15/17 : RUN chown -R $USER:$USER /home/$USER && chmod -R g-s /home/$USER && chmod -R o-wrx $WORKDIR && chmod o+rx /usr/local/bin/wait-for
 ---> Using cache
 ---> 5d26266f7dbb
Step 16/17 : USER $USER
 ---> Using cache
 ---> 93771cb45e5d
Step 17/17 : EXPOSE 4000
 ---> Using cache
 ---> 882d4e83840c
```

binarymist commented:
Yes the cache is working, but when someone makes a change to a source code file, then everything after that (the `COPY` of `node_modules` for example) will need to be rebuilt, taking additional time that would not happen if the commands were around the other way. Unless the first-stage build gets around this somehow? Easiest way is to just try making a change in one of the source files and see if the steps after `COPY . $WORKDIR` are being re-built.

mostafahussein commented:
If the code changed then the first stage will be triggered again (with new npm installation) and then the whole cache will be invalidated.

mostafahussein commented:
Good to go :+1:

binarymist commented:
Looking good @mostafahussein

binarymist commented:
@lirantal @ckarande : thoughts about trusting that the script at https://raw.githubusercontent.com/eficode/wait-for/master/wait-for will remain trust-worthy forever?

mostafahussein commented:
Take this into consideration:
If we cannot trust the script in long term, we can add it to `nodegoat` repository in a folder called `docker`. Currently the script has not been updated since 2017 and yes it might be updated in the feature but the basic functionality that we only need at the current moment is enough which makes us able to have a copy of it without worrying about feature updates unless there is a must.

lirantal commented on Dockerfile:
```diff
@@ -1,32 +1,21 @@
+COPY package*.json $WORKDIR
+RUN npm install --production --no-cache
+
+FROM node:4-alpine
```
the node image has its own built-in `node` user, perhaps we can use that instead of creating a special USER and GROUP for it?

see reference for the the dockerfile here: https://github.com/nodejs/docker-node/blob/master/Dockerfile-alpine.template

mostafahussein replied:
I thought about this too but during comparing node user with custom user I assumed that we might need to define our own uid/gid later and of course we need to disable the shell acess for the user. That's why I didn't mention the possibilty of using the existing node user instead of creating another

lirantal replied:
Even if that's the case, I don't think the scope of the project and/or container is to not allow users to login through the shell...

mostafahussein replied:
Okay then Its better to use the node user, should I wait for confirmtion from other reviewers or should I go and modify it ?

lirantal replied:
Let's wait and see what the other say

binarymist replied:
Yeah, I think your case to allow shell makes sense @lirantal

mostafahussein replied:
This one is done

lirantal commented on Dockerfile:
```diff
@@ -1,32 +1,21 @@
+RUN addgroup -S $GROUP && adduser -S -D -s /bin/false $USER -G $GROUP
+WORKDIR $WORKDIR
+COPY --from=0 /usr/src/app/node_modules node_modules
+ADD https://raw.githubusercontent.com/eficode/wait-for/master/wait-for /usr/local/bin/wait-for
```
@binarymist how about we use something like https://www.npmjs.com/package/wait-on ?
we can add it to the package manifest for nodegoat, or even better spawn it as a one off with `npx wait-on 'something'`

I understand we're taking a chance to trust on third-parties but there has to be some level of trust anyway, worst-case if one of those deps get deprecated we can just replace it with something else.

binarymist replied:
I like `npx`, but it removes an opportunity for the likes of snyk to analyse the package manifest (I think), and although it's kind of a slick way of running arbitary code, by the look of it, I don't think it's any better than adding and invoking https://raw.githubusercontent.com/eficode/wait-for/master/wait-for . Thoughts?

mostafahussein replied:
What about using the wait-for function itself as a standalone code inside `entrypoint.sh` and remove the wait-for usage ? its just a for loop anyway as you can see in here:
```
  for i in `seq $TIMEOUT` ; do
    nc -z "$HOST" "$PORT" > /dev/null 2>&1
    
    result=$?
    if [ $result -eq 0 ] ; then
      if [ $# -gt 0 ] ; then
        exec "$@"
      fi
      exit 0
    fi
    sleep 1
  done
  echo "Operation timed out" >&2
  exit 1
```

lirantal replied:
RE npx and `wait-on` - I don't think there's any expectation to scan those. They are helpers in building the docker image and having an environment setup, not part of the application itself.

I think it is more likely for the gist to go offline than the npm package to get removed.
If we still want to keep the bash script then I'd copy it (with whatever license and copyright notice it has) to our own repo and use as a helper through the build, without needing to hit github's gists for it.

mostafahussein replied:
So what is the last decision regarding this matter? Should we go with offline wait-for script ?

binarymist replied:
I'm not sure if @ckarande has snyk running on NodeGoat or not. I don't see any difference in running arbitrary code in an environment setup as opposed to running arbitrary code as part of the application. Either way, it's still arbitrary code execution on someone's system.

If it was my decision, I'd have something reputable checking for [A9](https://www.owasp.org/index.php/Top_10-2017_A9-Using_Components_with_Known_Vulnerabilities)'s, whether it was environmental or app code. I don't think this is my decision to make though. I'd like to see @ckarande make this decision known.

I think a npm package makes better sense than a script that's not in npm in terms of availability, but personally, I'd be happier still if the few lines were pulled into NodeGoat, but then if we have something like snyk checking the package manifest, then it's slightly less of an issue.

A decision from @ckarande is probably required here.

mostafahussein replied:
@lirantal @binarymist @ckarande I have removed wait-for and used a normal bash command instead, i think this resolves this conversation as we don't need to worry about using 3rd party script or npm module, I have tested it and works as expected

lirantal commented:
@binarymist @ckarande @mostafahussein added some comments

binarymist commented:
Tested locally, including shelling into the container. Thanks for your hard work on this @mostafahussein!
